keps/sig-cli/kubectl-socket-support.md

```diff
@@ -0,0 +1,184 @@
+
+## Proposal
+
+### User Stories
+
+To access a Kubernetes cluster, users will issue `kubectl` using an appropriate kubeconfig with scoped roles. However it is
+hard to avoid the kubeconfig files get spread, which will bring risks for production clusters. Moreover, it is not easy to revoke
+those kubeconfig files unless they get expired, especially TLS certificates.
+
+From the perspective of security, a more dynamic kubeconfig is preferred, such as using unix sockets, which could be revoked and closed
+as needed. As well this will bring good isolation when socket file is read-only by specific users.
+
+```bash
+export TOKEN=USER1_TOKEN
+kubectl proxy --unix-socket=/home/user1/.kube/kube.sock --token=${TOKEN}
+```
+
+Such of proxies could be setup automatically by administrators on demand. There is no need to expose tokens, certificates to users.
+
+Then user `user1` can issue below command to the cluster, while other users have no access to such socket file. It is a good way to keep it
+confidential.
+
+```bash
+kubectl --server=/home/user1/.kube/kube.sock get pods
+```
+
+### Implementation Details/Notes/Constraints
+
+Also socket addresses can be specified as `Cluster.Server`.
+
+```go
+// staging/src/k8s.io/client-go/tools/clientcmd/api/v1/types.go
+// Cluster contains information about how to communicate with a kubernetes cluster
+type Cluster struct {
+    // Server is the address of the kubernetes cluster (https://hostname:port).
+    // Socket address is still supported in experimental, such as unix:///var/run/kube.sock, tcp://192.168.59.106
+    Server string `json:"server"`
+    ...
+}
+```
+
+To connect to apiserver via sockets, we also need to modify current `TransportConfig` to convey unix sockets over HTTP.
+
+```go
+// staging/src/k8s.io/client-go/rest/transport.go
+// TransportConfig converts a client config to an appropriate transport config.
+func (c *Config) TransportConfig() (*transport.Config, error) {
+    conf := &transport.Config{
+    	...
+    	Dial: c.Dial,
+    }
+    
+    if scheme, sockpath, err := isSocketAddr(c.Host); err == nil {
+        conf.Dial = func(_ context.Context, _, _ string) (net.Conn, error) {
+            return net.DialTimeout(scheme, sockpath, defaultTimeout)
+        }
+    }   
+    ...
+}
+```
+
+### Risks and Mitigations
+
+There should be no compatibility issues. New clients will still be compatible with old servers and vice-versa.
+And it will bring no change for in-cluster accessing since such unix socket cannot be speficied
+by `KUBERNETES_SERVICE_HOST` and `KUBERNETES_SERVICE_PORT`.
+
+This proposal will need to modify `k8s.io/client-go` to allow `Cluster.Server` to be a unix socket. So technically unix sockets can be used
+by webhooks and operators as well. For cluster deployers and administrators, please be aware to keep unix socket file confidential by setting right file modes
+to avoid being misused.
```

Comment:
In an application that 1) accepts a kubeconfig file and 2) was separately expected to make use of sockets for other purposes, this new capability in a kubeconfig+client-go combination would allow anyone controlling the kubeconfig file to make requests to/from the socket, right? File modes/permissions would not be sufficient to prevent this, since the application is expected to use the socket, but the kubernetes client inside the application is not.